ambry-quota/src/main/java/com/github/ambry/quota/AmbryQuotaManager.java

```diff
@@ -77,86 +88,138 @@ public AmbryQuotaManager(QuotaConfig quotaConfig, QuotaRecommendationMergePolicy
   }
 
   @Override
-  public void init() throws InstantiationException {
-    Timer.Context timer = quotaMetrics.quotaManagerInitTime.time();
-    try {
-      for (QuotaEnforcer quotaEnforcer : requestQuotaEnforcers) {
-        quotaEnforcer.init();
-      }
-    } catch (Exception e) {
-      logger.error("Failed to init quotaEnforcer", e);
-      if (e instanceof InstantiationException) {
-        throw (InstantiationException) e;
-      } else {
-        throw new InstantiationException(e.getMessage());
+  public synchronized void init() throws Exception {
+    if (!initialized.get()) {
+      Timer.Context timer = quotaMetrics.quotaManagerInitTime.time();
+      try {
+        for (QuotaEnforcer quotaEnforcer : quotaEnforcers) {
+          quotaEnforcer.init();
+        }
+      } catch (Exception e) {
+        LOGGER.error("Failed to init quotaEnforcer", e);
+        throw e;
+      } finally {
+        timer.stop();
       }
-    } finally {
-      timer.stop();
+      LOGGER.info("AmbryQuotaManager initialized.");
+      initialized.set(true);
+    } else {
+      LOGGER.warn("AmbryQuotaManager already initialized.");
     }
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * If there are any {@link QuotaEnforcer#recommend} calls that throw exception, this implementation logs the exception,
+   * and makes a best effort to gather as many recommendations as possible. Recommendations for some {@link QuotaName}s
+   * will be missing in those cases.
+   */
   @Override
-  public ThrottlingRecommendation getThrottleRecommendation(RestRequest restRequest) {
-    if (requestQuotaEnforcers.isEmpty()) {
+  public ThrottlingRecommendation recommend(RestRequest restRequest) throws QuotaException {
+    quotaMetrics.recommendRate.mark();
+    if (quotaEnforcers.isEmpty()) {
+      LOGGER.warn("No quota enforcers in AmbryQuotaManager. Returning null recommendation.");
```

Comment:
let's use trace or debug here, but add a warn log in the constructor. This is probably not going to happen, but if it does, we are going to see tons of warn log messages in the file.

Reply:
done.